gh-127833: Docs: Add a `grammar-snippet` directive & replace `productionlist`

encukou opened the pull request:
As a first step to what's proposed in the issue, this overrides the ReST `productionlist` directive to:
- use `:` instead of the `::=` symbol
- add syntax highlighting for strings (the implementation is preliminary; I intend to later add a proper class to the theme instead of borrowing a Pygments class.)

All links and link targets should be preserved. (Unfortunately, this reaches into some Sphinx internals; I don't see a better way to do exactly what Sphinx does.)

This also adds a new directive, `grammar-snippet`, which formats the snippet almost exactly like what's in the source, modulo syntax highlighting and keeping the backtick character to mark links to other rules.
This will allow formatting the snippets as in the [grammar file](file:///home/encukou/dev/cpython/Doc/build/html/reference/grammar.html) (most importantly: to remove the the indentation that Sphinx adds to align the text).

I'd appreciate if a Sphinx expert could take a look to see if there's a better way to do something :)  cc @AA-Turner 

<!-- gh-issue-number: gh-127833 -->
* Issue: gh-127833
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--127835.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

encukou committed:
First version of GrammarSnippetDirective, copied from an earlier branch

Co-authored-by: Blaise Pabon <blaise@gmail.com>
Co-authored-by: William Ferreira <wqferr@gmail.com>

blaisep committed:
Add a grammar-snippet directive

blaisep committed:
Remove monkey patch for production list and Add CompatProductionList

to conserve the productionlist usage in the docs.

encukou committed:
Fix ReST grammar

encukou committed:
Complete the CompatProductionList class

Co-Authored-By: bswck <bartoszpiotrslawecki@gmail.com>

encukou committed:
Adjust docstrings & move GrammarSnippetDirective class up

encukou committed:
Run Ruff

encukou committed:
Merge in the main branch

encukou committed:
Run Ruff again with different flags

AA-Turner commented on Doc/reference/toplevel_components.rst:
```diff
@@ -66,7 +66,9 @@ File input
 All input read from non-interactive files has the same form:
 
-.. productionlist:: python-grammar
+.. grammar-snippet::
+   :group: python-grammar
```
Why `:group:` instead of a directive argument?

encukou replied:
Good question :)

I'd eventually like to use directive arguments for listing the rules to be documented (here, `file_input`), with the body of the directive being autogenerated or at least checked.
I could keep using arguments for group instead, but eventually I'd add *some* option, which means requiring an empty line.

The changed syntax also makes it easier to see if it's the new directive or the old one.

Does that make sense? I can change it back, of course.

AA-Turner replied:
Perhaps we could use that argument now? Even if unused, it shows what the intention is.

P.S. you probably don't want `final_argument_whitespace = True`, that's often for when we merge parsing of directive arguments and content (e.g. admonitions can be used with or without a blank line).

encukou replied:
(resolved "offline")

AA-Turner commented on Doc/tools/extensions/grammar_snippet.py:
```diff
@@ -0,0 +1,192 @@
+    optional_arguments = 1
+    final_argument_whitespace = True
+
+    def run(self):
```
Optionally, you could add type hints:

```suggestion
    def run(self) -> list[nodes.paragraph]:
```

encukou replied:
Is there any documentation on how to locally check if the hints are correct?
I tried running `mypy Doc/tools/extensions/` but got a bunch of unrelated errors.

AA-Turner replied:
I rely on PyCharm. I might add type checking once I'm done splitting pyspecific into individual extensions, but I don't know it it'll be worth it yet.

AA-Turner commented on Doc/tools/extensions/grammar_snippet.py:
```diff
@@ -0,0 +1,192 @@
+        return make_snippet(self, self.options, self.content)
+
+
+def make_snippet(directive, options, content):
```
Would you consider implementing `CompatProductionList` as a subclass of `GrammarSnippet` and overriding the relevent bits? It feels a bit odd to pass the full `directive` object to a free-standing object, though I understand why it is being done.

encukou replied:
Conceptually, `CompatProductionList` is not a `GrammarSnippet`, so it feels wrong to subclass. The option/argument handling should be entirely separate, I don't want settings from `GrammarSnippet` to become defaults for `CompatProductionList`.
So I added a common base.

AA-Turner commented on Doc/tools/extensions/grammar_snippet.py:
```diff
@@ -0,0 +1,192 @@
+                    )
+                    domain.note_object(
+                        'token', obj_name, node_id, location=name_node
+                    )
```
Could we factor this to a helper method? Is this code intended to be transitional or will it stay long-term?

encukou replied:
Sounds good!

AA-Turner commented on Doc/tools/extensions/grammar_snippet.py:
```diff
@@ -0,0 +1,192 @@
+                    ref_node += nodes.Text(name)
+                    literal += ref_node
+                case {'single_quoted': name} | {'double_quoted': name}:
+                    string_node = nodes.inline(classes=['nb'])
```
I assume `'nb'` is a Pygments class? Is there a way to avoid hard-coding this?

encukou replied:
Yes, it's a Pygments class.
This is a chicken-and-egg problem. My plan is to merge `nb`, then add a new class to python-docs-theme, and then go back here to use the new class.

AA-Turner replied:
OK. Another option is adding a small CSS file in `Doc/tools/static/` and using `app.add_css_file()` which we would then later merge in to p-d-t.

hugovk replied:
Or add it to p-d-t now and we can release it right away?

encukou replied:
I'll add the Pygments class for "other strings".

AA-Turner reviewed (commented):
A few initial notes. The targets problem is annoying, Sphinx doesn't make it easy...

encukou committed:
Merge in the main branch

encukou committed:
Turn the common function into a class with method; break make_link_to_token out

encukou committed:
Remove blank line

encukou committed:
Run `ruff format`

encukou committed:
Add forgotten arguments

encukou committed:
Merge in the main branch

encukou committed:
Use Sphinx's token_xrefs function for formatting the tokens

encukou committed:
Improve the highlight class injection

encukou committed:
Appease Ruff

encukou committed:
Appase Ruff again

AA-Turner committed:
Add type annotations, minor clean-ups

- Add a module docstring
- Use the snake_case convention for node classes
- Make :group: a required option
- declare parallel_write_safe = True

AA-Turner commented on Doc/tools/extensions/grammar_snippet.py:
```diff
@@ -138,20 +160,20 @@ class GrammarSnippetDirective(GrammarSnippetBase):
     has_content = True
     option_spec = {
-        'group': directives.unchanged,
+        'group': directives.unchanged_required,
```
This enforces that the `:group:` option is non-empty.

AA-Turner commented on Doc/tools/extensions/grammar_snippet.py:
```diff
@@ -7,21 +12,38 @@
 
-    def make_grammar_snippet(self, options, content):
+    def make_grammar_snippet(
+        self, options: dict[str, Any], content: Sequence[str]
```
Currently `options` is `str -> str` as the only key is `group`, but future options may have different allowed values.

AA-Turner commented on Doc/tools/extensions/grammar_snippet.py:
```diff
@@ -7,21 +12,38 @@
     """Node for a string literal in a grammar snippet."""
 
-    def __init__(self) -> None:
+    def __init__(
+        self,
+        rawsource: str = '',
+        text: str = '',
+        *children: Node,
+        **attributes: Any,
+    ) -> None:
```
To match the constructor arguments of the superclass (this also allows for a small simplification later on)

AA-Turner reviewed (approved):
I'm happy with this as it stands, thank you @encukou for the discussion. As discussed, I've pushed a commit to add type annotations. I've included some other minor changes and left a self-review for a brief explanation -- feel free to revert any you disagree with.

A
